lib/src/command/add.dart

```diff
@@ -0,0 +1,339 @@
+    /// Perform version resolution in-memory.
+    final updatedPubSpec =
+        await _addPackageToPubspec(entrypoint.root.pubspec, package);
+    final solveResult = await resolveVersions(
+        SolveType.GET, cache, Package.inMemory(updatedPubSpec));
+
+    final resultPackage = solveResult.packages
+        .firstWhere((packageId) => packageId.name == package.name);
+
+    /// Assert that [resultPackage] is within the original user's expectations.
+    if (package.constraint != null &&
+        !package.constraint.allows(resultPackage.version)) {
+      dataError('${package.name} resolved to ${resultPackage.version} which '
+          'does not match the input ${package.constraint}! Exiting.');
+    }
+
+    if (isDryRun) {
+      /// Even if it is a dry run, run `acquireDependencies` so that the user
+      /// gets a report on the other packages that might change version due
+      /// to this new dependency.
+      final newRoot = Package.inMemory(updatedPubSpec);
+
+      await Entrypoint.global(newRoot, entrypoint.lockFile, cache)
+          .acquireDependencies(SolveType.GET,
+              dryRun: true, precompile: argResults['precompile']);
+    } else {
+      /// Update the pubspec. It is necessary to do this before calling
+      /// [acquireDependencies] (as opposed to calling [acquireDependencies] on
+      /// the in-memory pubspec and writing it later) because there will be
+      /// conflicts in other parts of [pub] if the pubspec is newer than
+      /// `pubspec.lock`.
+      _updatePubspec(resultPackage, packageInformation, isDev);
+
+      /// Create a new [Entrypoint] since we have to reprocess the updated
+      /// pubspec file.
+      await Entrypoint.current(cache).acquireDependencies(SolveType.GET,
+          precompile: argResults['precompile']);
+    }
+
+    if (isOffline) {
+      log.warning('Warning: Packages added when offline may not resolve to '
+          'the latest versions of your dependencies.');
+    }
+  }
+
+  /// Creates a new in-memory [Pubspec] by adding [package] to the
+  /// dependencies of [original].
+  Future<Pubspec> _addPackageToPubspec(
+      Pubspec original, PackageRange package) async {
+    ArgumentError.checkNotNull(original, 'original');
+    ArgumentError.checkNotNull(package, 'package');
+
+    final dependencies = [...original.dependencies.values];
+    var devDependencies = [...original.devDependencies.values];
+
+    if (isDev) {
+      final dependencyNames = dependencies.map((dependency) => dependency.name);
+
+      /// If package is originally in dependencies and we wish to add it to
+      /// dev_dependencies, this is a redundant change, and we should not
+      /// remove the package from dependencies, since it might cause the user's
+      /// code to break.
+      if (dependencyNames.contains(package.name)) {
+        usageException('${package.name} is already in dependencies. '
+            'Please remove existing entry before adding it to dev_dependencies');
```

Comment:
Something like  `Use 'pub remove ${package.name} to remove it before adding it to 'dev_dependencies'`